### PR TITLE
Fix bug in rollback txn case

### PR DESCRIPTION
**Pull request by yiguolei**
1. Move clear alter task logic to a common method
2. Rollback the txn when batch load failed

### CODE REVIEW

File: be/src/olap/txn_manager.cpp [hunk 1]
```diff
@@ -263,6 +266,10 @@ OLAPStatus TxnManager::publish_txn(OlapMeta* meta, TPartitionId partition_id, TT
     }
 }
 
+// txn could be rollbacked if it does not have related rowset
+// if the txn has related rowset then could not rollback it, because it 
+// may be committed in another thread and our current thread meets erros when writing to data file
+// be has to wait for fe call clear txn api
```
Comment (chaoyli) [inline be/src/olap/txn_manager.cpp]: In what scenario, one thread want to rollback and another thread commit it?

File: be/src/olap/delta_writer.cpp [hunk 2]
```diff
@@ -46,12 +46,20 @@ DeltaWriter::~DeltaWriter() {
 }
 
 void DeltaWriter::_garbage_collection() {
-    TxnManager::instance()->rollback_txn(_req.partition_id, _req.txn_id,_req.tablet_id, _req.schema_hash);
-    StorageEngine::instance()->add_unused_rowset(_cur_rowset);
+    OLAPStatus rollback_status = TxnManager::instance()->rollback_txn(_req.partition_id,
```
Comment (chaoyli) [inline be/src/olap/delta_writer.cpp]: In this place, if data has generated but rowset is not been created, how to GC data?

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/olap/txn_manager.cpp (hunk 1 above)
Comment (yiguolei) [reply be/src/olap/txn_manager.cpp]: In stream load scenario, if coordinator be send two ingest commands, there maybe two running thread to load the rowset.

File: be/src/olap/tablet.cpp [hunk 3]
```diff
@@ -635,6 +637,24 @@ OLAPStatus Tablet::delete_alter_task() {
     return _tablet_meta->delete_alter_task();
 }
 
+OLAPStatus Tablet::protected_delete_alter_task() {
```
[the remark this answers is not in the document]
Comment (yiguolei) [reply be/src/olap/tablet.cpp]: No. sometimes delete alter task is called under a write lock, and call other methods like _drop_tablet_directly_unlocked, so could not merge them.

File: be/src/olap/task/engine_clear_alter_task.cpp [hunk 4]
```diff
@@ -46,26 +46,21 @@ OLAPStatus EngineClearAlterTask::_clear_alter_task(const TTabletId tablet_id,
...
-    tablet->delete_alter_task();
-    OLAPStatus res = tablet->save_tablet_meta();
-    if (res != OLAP_SUCCESS) {
-        LOG(FATAL) << "fail to save header. [res=" << res << " tablet='" << tablet->full_name() << "']";
-    } else {
-        LOG(INFO) << "clear alter task on tablet. [tablet='" << tablet->full_name() << "']";
-    }
-    tablet->release_header_lock();
+    OLAPStatus res = OLAP_SUCCESS;
+    res = tablet->protected_delete_alter_task();
```
[the remark this answers is not in the document]
Comment (yiguolei) [reply be/src/olap/task/engine_clear_alter_task.cpp]: done

**Pull request merged**